macros/parserPopUp.pl

```diff
@@ -90,51 +121,73 @@ sub new {
   my $self = shift; my $class = ref($self) || $self;
   shift if Value::isContext($_[0]); # remove context, if given (it is not used)
   my $choices = shift; my $value = shift;
-  Value::Error("A PopUp's first argument should be a list of menu items")
+  Value->Error("A PopUp's first argument should be a list of menu items")
     unless ref($choices) eq 'ARRAY';
-  Value::Error("A PopUp's second argument should be the correct menu choice")
+  Value->Error("A PopUp's second argument should be the correct menu choice")
     unless defined($value) && $value ne "";
-  my %choice; map {$choice{$_} = 1} @$choices;
-  Value::Error("The correct choice must be one of the PopUp menu items")
-    unless $choice{$value};
   $self = bless {data => [$value], context => $context, choices => $choices}, $class;
+  $self->getChoiceOrder;
+  my %choice; map {$choice{$_} = 1} @{$self->{choices}};
+  if (!$choice{$value}) {
+    my @order = map {ref($_) eq "ARRAY" ? @$_ : $_} @$choices;
+    if ($value =~ m/^\d+$/ && $order[$value]) {$self->{data}[0] = $order[$value]}
+      else {Value->Error("The correct choice must be one of the PopUp menu items")}
+  }
   return $self;
 }
 
 #
+#  Get the choices into the correct order (randomizing where requested)
+#
+sub getChoiceOrder {
+  my $self = shift;
+  my @choices = ();
+  foreach my $choice (@{$self->{choices}}) {
+    if (ref($choice) eq "ARRAY") {push(@choices,$self->randomOrder($choice))}
+      else {push(@choices,$choice)}
+  }
+  $self->{choices} = \@choices;
+}
+sub randomOrder {
+  my $self = shift; my $choices = shift;
+  my %index = (map {$main::PG_random_generator->rand => $_} (0..scalar(@$choices)-1));
+  return (map {$choices->[$index{$_}]} main::PGsort(sub {$_[0] lt $_[1]},keys %index));
+}
+
+#
 #  Create the menu list
 #
 sub menu {shift->MENU(0,@_)}
 sub MENU {
   my $self = shift; my $extend = shift; my $name = shift;
-  my $list = $self->{choices}; my $menu = "";
+  my @list = @{$self->{choices}}; my $menu = "";
   $name = main::NEW_ANS_NAME() unless $name;
   my $answer_value = (defined($main::inputs_ref->{$name}) ? $main::inputs_ref->{$name} : '');
   my $label = main::generate_aria_label($name);
   if ($main::displayMode =~ m/^HTML/) {
-    $menu = qq!<SELECT class="pg-select" NAME="$name" id="$name" aria-label="$label" SIZE=1>\n!;
-    foreach my $option (@$list) {
-      my $selected = ($option eq $answer_value) ? " SELECTED" : "";
-      $menu .= qq!<OPTION$selected VALUE="$option">$option</OPTION>\n!;
+    $menu = qq!<select class="pg-select" name="$name" id="$name" aria-label="$label" size="1" style="max-width:100%">\n!;
```

Comment:
Max width was 95% in math4 because without a max width input boxes and their like were spilling over the problem well on small screens.  Changing it to 95% made it so that even if the screen is small there is always a little boundary between the input elements and the problem well.   If there is a better way to acomplish that, I'm happy to change it.  I don't think it hurts anything to specify 100% here, other than the general principle that explicitly setting css makes it harder for things like themes to modify the look of page elements.